File: fetchers/cross_market.py
"""cross_market.py — Cross-platform prediction market price comparison

Searches Polymarket and Metaculus for markets matching a query, returns
current prices for arbitrage detection.

Usage (callable by agents via Bash):
    python -m fetchers.cross_market --query "government shutdown"
    python -m fetchers.cross_market --query "Fed rate cut" --platforms polymarket,metaculus

Output: JSON array of { platform, title, probability, volume, url }
"""

import sys
import json
import os
import argparse
from dotenv import load_dotenv

from src.config import (
    POLYMARKET_API_BASE, POLYMARKET_REQUEST_TIMEOUT, POLYMARKET_SEARCH_LIMIT,
    METACULUS_API_BASE, METACULUS_REQUEST_TIMEOUT, METACULUS_SEARCH_LIMIT,
)
from src import http_client

load_dotenv()


def search_polymarket(query: str, limit: int = POLYMARKET_SEARCH_LIMIT) -> list[dict]:
    """Search Polymarket for matching markets. No auth required."""
    resp = http_client.get(
        f"{POLYMARKET_API_BASE}/public-search",
        params={"q": query, "limit_per_type": limit},
        timeout=POLYMARKET_REQUEST_TIMEOUT,
    )
    if not resp:
        return []

    data = resp.json()
    results = []
    for event in data.get("events", []):
        for market in event.get("markets", []):
            if market.get("closed"):
                continue
            try:
                prices = json.loads(market.get("outcomePrices", "[]"))
                yes_prob = float(prices[0]) if prices else None
            except (json.JSONDecodeError, IndexError, ValueError):
                yes_prob = None
            if yes_prob is None:
                continue

            results.append({
                "platform": "polymarket",
                "title": market.get("question") or market.get("groupItemTitle", ""),
                "probability": round(yes_prob, 4),
                "volume_usd": round(float(market.get("volume", 0)), 2),
                "url": f"https://polymarket.com/event/{event.get('slug', '')}",
            })
    return results


def search_metaculus(query: str, limit: int = METACULUS_SEARCH_LIMIT) -> list[dict]:
    """Search Metaculus for matching questions. Requires METACULUS_TOKEN env var."""
    token = os.getenv("METACULUS_TOKEN")
    if not token:
        print("METACULUS_TOKEN not set — skipping Metaculus", file=sys.stderr)
        return []

    resp = http_client.get(
        f"{METACULUS_API_BASE}/posts/",
        params={
            "search": query,
            "statuses": "open",
            "forecast_type": "binary",
            "limit": limit,
            "order_by": "-hotness",
            "with_cp": "true",
        },
        headers={"Authorization": f"Token {token}"},
        timeout=METACULUS_REQUEST_TIMEOUT,
    )
    if not resp:
        return []

    results = []
    for post in resp.json().get("results", []):
        question = post.get("question") or {}
        aggregations = question.get("aggregations", {})

        prob = None
        for agg_type in ("recency_weighted", "unweighted"):
            agg = aggregations.get(agg_type, {})
            latest = agg.get("latest", {})
            centers = latest.get("centers", [])
            if centers:
                prob = centers[0]
                break

        if prob is None:
            continue

        post_id = post.get("id", "")
        results.append({
            "platform": "metaculus",
            "title": post.get("title", ""),
            "probability": round(float(prob), 4),
            "forecasters": post.get("nr_forecasters", 0),
            "url": f"https://www.metaculus.com/questions/{post_id}/",
        })
    return results


PLATFORM_FUNCS = {
    "polymarket": search_polymarket,
    "metaculus": search_metaculus,
}


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Cross-market price comparison")
    parser.add_argument("--query", required=True, help="Search query (market topic)")
    parser.add_argument("--platforms", default="polymarket,metaculus",
                        help="Comma-separated platforms (default: polymarket,metaculus)")
    parser.add_argument("--limit", type=int, default=5, help="Max results per platform")
    args = parser.parse_args()

    platforms = [p.strip() for p in args.platforms.split(",")]
    all_results = []
    for platform in platforms:
        func = PLATFORM_FUNCS.get(platform)
        if not func:
            print(f"Unknown platform: {platform}", file=sys.stderr)
            continue
        all_results.extend(func(args.query, args.limit))

    sys.stdout.buffer.write((json.dumps(all_results, indent=2) + "\n").encode("utf-8"))

File: lib/fetchers/crossMarket.ts
import { getJson } from "../httpClient";
import {
  POLYMARKET_API_BASE,
  POLYMARKET_REQUEST_TIMEOUT,
  POLYMARKET_SEARCH_LIMIT,
  METACULUS_API_BASE,
  METACULUS_REQUEST_TIMEOUT,
  METACULUS_SEARCH_LIMIT,
} from "../config";
import type { CrossMarketResult } from "../types";

export async function searchPolymarket(
  query: string,
  limit = POLYMARKET_SEARCH_LIMIT,
): Promise<CrossMarketResult[]> {
  const data = await getJson<Record<string, unknown[]>>(
    `${POLYMARKET_API_BASE}/public-search`,
    {
      params: { q: query, limit_per_type: String(limit) },
      timeout: POLYMARKET_REQUEST_TIMEOUT,
    },
  );
  if (!data) return [];

  const results: CrossMarketResult[] = [];
  const events = (data.events || []) as Record<string, unknown>[];

  for (const event of events) {
    const markets = (event.markets || []) as Record<string, unknown>[];
    for (const m of markets) {
      if (m.closed === true || m.archived === true) continue;
      let probability = 0;
      try {
        const prices = JSON.parse((m.outcomePrices as string) || "[]") as string[];
        probability = parseFloat(prices[0] || "0");
      } catch {
        continue;
      }
      results.push({
        platform: "polymarket",
        title: (m.question as string) || (event.title as string) || "",
        probability,
        volume: m.volume_num_24hr as number | undefined,
        url: `https://polymarket.com/event/${event.slug}`,
      });
    }
  }

  return results.slice(0, limit);
}

export async function searchMetaculus(
  query: string,
  limit = METACULUS_SEARCH_LIMIT,
): Promise<CrossMarketResult[]> {
  const token = process.env.METACULUS_TOKEN;
  if (!token) return [];

  const data = await getJson<{ results: Record<string, unknown>[] }>(
    `${METACULUS_API_BASE}/posts/`,
    {
      params: {
        search: query,
        statuses: "open",
        forecast_type: "binary",
        limit: String(limit),
        order_by: "-hotness",
        with_cp: "true",
      },
      headers: { Authorization: `Token ${token}` },
      timeout: METACULUS_REQUEST_TIMEOUT,
    },
  );
  if (!data?.results) return [];

  const results: CrossMarketResult[] = [];
  for (const post of data.results) {
    const q = post.question as Record<string, unknown> | undefined;
    if (!q) continue;

    let probability = 0;
    try {
      const agg = q.aggregations as Record<string, unknown>;
      const rw = agg?.recency_weighted as Record<string, unknown>;
      const latest = rw?.latest as Record<string, unknown>;
      const centers = latest?.centers as number[];
      probability = centers?.[0] || 0;
    } catch {
      continue;
    }

    results.push({
      platform: "metaculus",
      title: (post.title as string) || "",
      probability,
      forecasters: q.forecasters_count as number | undefined,
      url: `https://www.metaculus.com/questions/${post.id}/`,
    });
  }

  return results;
}

export async function searchCrossMarket(
  query: string,
  platforms: string[] = ["polymarket", "metaculus"],
  limit = 5,
): Promise<CrossMarketResult[]> {
  const promises: Promise<CrossMarketResult[]>[] = [];
  if (platforms.includes("polymarket")) {
    promises.push(searchPolymarket(query, limit));
  }
  if (platforms.includes("metaculus")) {
    promises.push(searchMetaculus(query, limit));
  }
  const arrays = await Promise.all(promises);
  return arrays.flat();
}
